File: src/main/java/me/muksc/tacztweaks/mixin/accessor/EntityKineticBulletAccessor.java
package me.muksc.tacztweaks.mixin.accessor;

import com.tacz.guns.entity.EntityKineticBullet;
import com.tacz.guns.util.TacHitResult;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.gen.Accessor;
import org.spongepowered.asm.mixin.gen.Invoker;

@Mixin(value = EntityKineticBullet.class, remap = false)
public interface EntityKineticBulletAccessor {
    @Accessor("pierce")
    int tacztweaks$getPierce();

    @Accessor("pierce")
    void tacztweaks$setPierce(int pierce);

    @Accessor("explosion")
    boolean tacztweaks$getExplosion();

    @Invoker("onHitEntity")
    void tacztweaks$invokeOnHitEntity(TacHitResult result, Vec3 startVec, Vec3 endVec);

    @Invoker("onHitBlock")
    void tacztweaks$invokeOnHitBlock(BlockHitResult result, Vec3 startVec, Vec3 endVec);

}

File: src/main/kotlin/me/muksc/tacztweaks/mixininterop/EntityKineticBulletAccessor.kt
@file:Suppress("NOTHING_TO_INLINE")

package me.muksc.tacztweaks.mixininterop

import com.tacz.guns.util.TacHitResult
import me.muksc.tacztweaks.mixin.accessor.EntityKineticBulletAccessor
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3

inline var EntityKineticBulletAccessor.pierce: Int
    get() = `tacztweaks$getPierce`()
    set(value) = `tacztweaks$setPierce`(value)

inline val EntityKineticBulletAccessor.explosion: Boolean
    get() = `tacztweaks$getExplosion`()

inline fun EntityKineticBulletAccessor.onHitEntity(result: TacHitResult, startVec: Vec3, endVec: Vec3) =
    `tacztweaks$invokeOnHitEntity`(result, startVec, endVec)

inline fun EntityKineticBulletAccessor.onHitBlock(result: BlockHitResult, startVec: Vec3, endVec: Vec3) =
    `tacztweaks$invokeOnHitBlock`(result, startVec, endVec)